wandb/sdk/artifacts/artifact.py

```diff
@@ -1590,23 +1622,48 @@ def _download_file(
             _thread_local_api_settings.cookies = cookies
             _thread_local_api_settings.headers = headers
 
-            self.get_path(name).download(root)
+            entry.download(root)
             download_logger.notify_downloaded()
 
-        pool = multiprocessing.dummy.Pool(32)
-        pool.map(
-            partial(
-                _download_file,
-                api_key=_thread_local_api_settings.api_key,
-                cookies=_thread_local_api_settings.cookies,
-                headers=_thread_local_api_settings.headers,
-            ),
-            self.manifest.entries,
+        download_entry = partial(
+            _download_entry,
+            api_key=_thread_local_api_settings.api_key,
+            cookies=_thread_local_api_settings.cookies,
+            headers=_thread_local_api_settings.headers,
         )
+
+        with concurrent.futures.ThreadPoolExecutor(64) as executor:
+            active_futures = set()
+            # Download files.
+            has_next_page = True
+            cursor = None
+            while has_next_page:
+                attrs = fetch_file_urls(cursor)
+                has_next_page = attrs["pageInfo"]["hasNextPage"]
+                cursor = attrs["pageInfo"]["endCursor"]
+                for edge in attrs["edges"]:
+                    entry = self.get_path(edge["node"]["name"])
+                    entry._download_url = edge["node"]["directUrl"]
+                    active_futures.add(executor.submit(download_entry, entry))
+                # Wait for download threads to catch up.
+                max_backlog = 5000
```

Comment:
I'm a little worried that this could exceed the default 4096 ulimit that some machines have. It shouldn't, because that gets handled at a lower level, but I suspect that 1000 would still be performant. 5000 also seems much more likely to hit the 1-hour expiration; for mid to large size files on slow machines it almost certainly will.

Reply:
There are 64 threads downloading files. The main thread fetches URLs to download and puts them on a queue for the download threads to download. If the queue is longer than max backlog, we wait before fetching the next batch of URLs.

There are 3 parameters here:
- the number of download threads, currently 64
  - setting this too low means the CPU will be idle waiting for I/O
  - setting this too high means we open many files (2 per thread when copying the downloaded file to cache)
- the batch size, currently 5k
  - setting this too low means download threads will become idle waiting for the next batch, because even small batches are slow to generate since we have to download the manifest
  - setting this too high means we risk hitting a timeout on the server (the query takes ~23s for my 1M-file artifact, ~8s for fetching the manifest and ~3s / 1k URLs generated; the timeout is 60s) and having the URLs expire before we start downloading them
- max backlog, currently 5k
  - setting this too low means the download threads will become idle waiting for the next batch
  - setting this too high means the URLs might expire before we start downloading them